test/unit/common/test_shardtrie.py

```diff
@@ -0,0 +1,312 @@
+        """Test deleting a node from shardtrie"""
+        trie = self.trie
+        number_data_nodes = trie.root.count_data_nodes()
+
+        # Delete a non data node
+        self.assertTrue(trie.delete('ab'))
+        self.assertEqual(trie.root.count_data_nodes(), number_data_nodes)
+
+        # Delete a data node
+        self.assertTrue(trie.delete('abc'))
+        self.assertEqual(trie.root.count_data_nodes(), number_data_nodes - 1)
+
+    def test_shardtrie_add(self):
+        """Test adding a node to the shardtrie"""
+        trie = self.trie
+        number_data_nodes = trie.root.count_data_nodes()
+
+        self.assertTrue(trie.add('abce'))
+        self.assertEqual(trie.root.count_data_nodes(), number_data_nodes + 1)
+
+    def test_shardtrie_add_distributed(self):
+        """Test adding a node to a distributed shardtrie"""
+        trie = self.trie
+
+        trie.split_trie('ab')
+        with self.assertRaises(shardtrie.ShardTrieDistributedBranchException):
+            trie.add('abce')
+
+    def test_shardtrie_get_important_nodes(self):
+        """Test iterating across data nodes"""
+        trie = self.trie
+
+        # Iterate over all data nodes
+        self.assertTrue(
+            all([x.data != None for x in trie.get_important_nodes()]))
+
+        # Iterate over only 2 data nodes
+        self.assertEqual(
+            len(list(trie.get_important_nodes(limit = 2))),
+            2)
+
+        # Iterate starting after node abc
+        marker = trie.get_node('abc')
+        self.assertEqual(
+            list(trie.get_important_nodes(marker=marker))[0].full_key(),
+            'abd')
+
+        # Iterate starting at 'abc'
+        self.assertEqual(
+            list(trie.get_important_nodes('abc'))[0].full_key(),
+            'abc')
+
+    def test_shardtrie_dump(self):
+        """Test adding a node to a distributed shardtrie"""
+        trie = self.trie
+        self.assertEqual(trie.dump(), json.loads(self._test_shard_trie))
+
+    def test_shardtrie_get_last_node(self):
+        """Test getting the last node in the shardtrie"""
+        trie = self.trie
+        last_node = trie.get_last_node()
+        self.assertEqual(last_node.key, 'd')
+
+    def test_shardtrie_get_distributed_nodes(self):
+        """Test getting distributed nodes after shard"""
+        trie = self.trie
+
+        distributed_nodes = trie.get_distributed_nodes()
+        self.assertEqual(len(list(distributed_nodes)), 0)
+
+        subtrie = trie.split_trie('ab')
+
+        distributed_nodes = trie.get_distributed_nodes()
+        self.assertEqual(len(list(distributed_nodes)), 1)
+
+    def test_shardtrie_get_large_subtries(self):
+        """Test getting subtries over certain size"""
+        trie = self.trie
+
+        for i in range(20):
+            trie.add('abc%d' % i)
+            trie.add('abd%d' % i)
+
+        large_tries = [x[2] for x in trie.get_large_subtries(15)]
+        self.assertEqual(large_tries, ['abd', 'abc', 'ab'])
```

Comment:
This is the original way of getting the largest subtries.. it works so I say we leave it in for now. This uses way too much memory when dealing with _very_ large containers, and hence gave birth to the counting trie technique.